agent/agent.go

```diff
@@ -3465,4 +3577,142 @@ func (a *Agent) registerCache() {
 		RefreshTimer:   0 * time.Second,
 		RefreshTimeout: 10 * time.Minute,
 	})
+
+	a.cache.RegisterType(cachetype.ServiceHTTPChecksName, &cachetype.ServiceHTTPChecks{
+		Agent: a,
+	}, &cache.RegisterOptions{
+		Refresh:        true,
+		RefreshTimer:   0 * time.Second,
+		RefreshTimeout: 10 * time.Minute,
+	})
+}
+
+func (a *Agent) LocalState() *local.State {
+	return a.State
+}
+
+// rerouteExposedChecks will inject proxy address into check targets
+// Future calls to check() will dial the proxy listener
+// The agent stateLock MUST be held for this to be called
+func (a *Agent) rerouteExposedChecks(serviceID string, proxyAddr string) error {
+	for _, c := range a.checkHTTPs {
+		if c.ServiceID != serviceID {
+			continue
+		}
+		port, err := a.listenerPort(serviceID, string(c.CheckID))
+		if err != nil {
+			return err
+		}
+		addr := httpInjectAddr(c.HTTP, proxyAddr, port)
+		if err != nil {
+			// The only way to get here is if the regex pattern fails to compile, which would be caught by tests
+			return fmt.Errorf("failed to inject proxy addr into HTTP target")
+		}
+		c.ProxyHTTP = addr
+	}
+	for _, c := range a.checkGRPCs {
+		if c.ServiceID != serviceID {
+			continue
+		}
+		port, err := a.listenerPort(serviceID, string(c.CheckID))
+		if err != nil {
+			return err
+		}
+		addr := grpcInjectAddr(c.GRPC, proxyAddr, port)
+		c.ProxyGRPC = addr
+	}
+	return nil
+}
+
+// resetExposedChecks will set Proxy addr in HTTP checks to empty string
+// Future calls to check() will use the original target c.HTTP or c.GRPC
+// The agent stateLock MUST be held for this to be called
+func (a *Agent) resetExposedChecks(serviceID string) {
+	for _, c := range a.checkHTTPs {
+		if c.ServiceID == serviceID {
+			c.ProxyHTTP = ""
+		}
+	}
+	for _, c := range a.checkGRPCs {
+		if c.ServiceID == serviceID {
+			c.ProxyGRPC = ""
+		}
+	}
+	for k, _ := range a.exposedPorts {
+		if strings.HasPrefix(k, serviceID) {
+			delete(a.exposedPorts, k)
+		}
+	}
+}
+
+// listenerPort allocates a port from the configured range
+// The agent stateLock MUST be held when this is called
+func (a *Agent) listenerPort(svcID, checkID string) (int, error) {
```

Comment:
```suggestion
func (a *Agent) listenerPortLocked(svcID, checkID string) (int, error) {
```

appending `Locked` is the code in consul signals that locking is required by the caller.